ada_feeding/ada_feeding/decorators/force_status.py

```diff
@@ -0,0 +1,68 @@
+"""
+This module defines the ForceStatus decorator, which ignores the child's
+current status and always returns the status specified by the user.
+"""
+
+# Standard imports
+import typing
+
+# Third-party imports
+from py_trees import behaviour, common
+from py_trees.decorators import Decorator
+
+# Local imports
+
+
+class ForceStatus(Decorator):
+    """
+    This decorator ignores the child's current status and always returns
+    the status specified by the user. This can be useful e.g., to force a
+    composite without memory to move on to the next child while still ticking
+    the child of this decorator.
+
+    While this decorator's behavior can be achieved with a chain of "X is Y"
+    style decorators, it is often conceptually easier to reason about a decorator
+    always returning one status, as opposed to a chain of decorators that each
+    flip their child's status.
+    """
+
+    def __init__(self, name: str, child: behaviour.Behaviour, status: common.Status):
+        """
+        Initialise the decorator.
+
+        Args:
+            name: the decorator name
+            child: the child to be decorated
+            status: the status to return on :meth:`update`
+        """
+        super().__init__(name=name, child=child)
+        self.const_status = status
+
+    def tick(self) -> typing.Iterator[behaviour.Behaviour]:
+        """
+        Don't stop the child even if this decorator's status is non-RUNNING.
+
+        Yields:
+            a reference to itself
+        """
+        self.logger.debug(f"{self.__class__.__name__}.tick()")
+        # initialise just like other behaviours/composites
+        if self.status != common.Status.RUNNING:
+            self.initialise()
+        # interrupt proceedings and process the child node
+        # (including any children it may have as well)
+        for node in self.decorated.tick():
+            yield node
+        # resume normal proceedings for a Behaviour's tick
+        self.status = self.update()
```

Comment:
There are two reasons for this.

1. By default, [if a decorator returns a status other than RUNNING, it will call `stop(INVALID)` on all its children that are RUNNING](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/decorators.py#L140). So therefore, we can't have a decorator that returns something like FAILURE while still allowing its kids to be ticked (as is the desired behavior here).
2. Why is it desired behavior to have this decorator return something like FAILURE while still allowing its kids to be ticked? That has to do with how composites progress in `py_trees`.
    1. We need the "main" subtree (the subtree that contains `on_success` and `on_failure`) to exist *before* the subtree containing `on_preempt`, in order to get the right preemption order. 
    2. However, both [Selectors](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/composites.py#L455) and [Sequences](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/composites.py#L571C23-L571C23), regardless of whether they have memory or not, don't tick later children if earlier children return `RUNNING`. Therefore, if `ForceStatus` returns `RUNNING`, then the `on_preempt` subtree will not get ticked.
    3. Why is it important to for the `on_preempt` subtree to get ticked? Because throughout the `py_trees` library, `stop(INVALID)` is only called on children who have a status other than `INVALID`. Therefore, if the `on_preempt` never gets ticked while the "main" tree is running, then the preemption behavior won't run when the tree gets preempted.
    4. So that means that if our top-level node is a `Selector` we must use `ForceStatus(FAILURE)` to ensure the `on_preempt` subtree gets ticked, and if our top-level node is a `Sequence` we must use `ForceStatus(SUCCESS)` to ensure the `on_preempt` subtree gets ticked. Between the two, `Selector` is better because it [does not call `stop(INVALID)` on children on initialization](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/composites.py#L409) (whereas [Sequence does](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/composites.py#L540)), which would cause `on_preempt` to run.
    5. The final option for the top-level composite to be `Parallel`, because a `Parallel` is the only composite that can have multiple children that are `RUNNING` at once. The main challenge with a `Parallel` is that once it reaches a terminal status, [it will call `stop(INVALID)`  on all its `RUNNING` children](https://github.com/splintered-reality/py_trees/blob/0d5b39f2f6333c504406d8a63052c456c6bd1ce5/py_trees/composites.py#L778), which means that the `on_preempt` behavior will be run even if the "main" tree succeeds, which is undesirable.
        1. This brings up a final point about why, at least for `scoped_behavior`, we don't just use a Parallel with `OnPreempt` (e.g., [as `py_trees` does with `eventually`](https://github.com/splintered-reality/py_trees/pull/427)). While that will achieve the desired performance of `Post` being run on succeess, failure, or preemption, it will *always run in synchronously in `terminate`*. However, desired behavior is for it to run asynchronously in `update` on success/failure, and only synchronously in `terminate` on preemption.

I think that fully covers the nuances of `py_trees` that resulted in the final architecture for `eventually_swiss`/`scoped_behavior`.

Reply:
Let's talk about OnPreempt, since it seems like the non-BT-esque behavior of ForceStatus is designed to cater to the weird behavior of that.